compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java

```diff
@@ -449,29 +451,55 @@ public Variable emitForeignCall(ForeignCallLinkage linkage, LIRFrameState frameS
 
     @Override
     public void emitStrategySwitch(JavaConstant[] keyConstants, double[] keyProbabilities, LabelRef[] keyTargets, LabelRef defaultTarget, Variable value) {
+
         int keyCount = keyConstants.length;
         SwitchStrategy strategy = SwitchStrategy.getBestStrategy(keyProbabilities, keyConstants, keyTargets);
+
+        double minDensity = 1 / Math.sqrt(strategy.getAverageEffort());
+
+        Optional<Hasher> hasher = hasherFor(keyConstants, minDensity);
+
+        double hashTableSwitchDensity = hasher.map(h -> keyCount / (double) h.cardinality()).orElse(0d);
+
         long valueRange = keyConstants[keyCount - 1].asLong() - keyConstants[0].asLong() + 1;
         double tableSwitchDensity = keyCount / (double) valueRange;
+
         /*
          * This heuristic tries to find a compromise between the effort for the best switch strategy
          * and the density of a tableswitch. If the effort for the strategy is at least 4, then a
          * tableswitch is preferred if better than a certain value that starts at 0.5 and lowers
          * gradually with additional effort.
          */
-        if (strategy.getAverageEffort() < 4 || tableSwitchDensity < (1 / Math.sqrt(strategy.getAverageEffort()))) {
+        if (strategy.getAverageEffort() < 4d || (tableSwitchDensity < minDensity && hashTableSwitchDensity < minDensity)) {
```

Comment:
Shouldn't this be || instead of &&?  If the hashTableSwitchDensity is good but tableSwitchDensity isn't wouldn't we still want to use the hash?

Reply:
The condition seems correct: 
- if both tables have density lower than `minDensity`, use the strategy switch
- if one of them is higher than `minDensity` , use one of the table switches

Reply:
Yes of course you're right.  I was just misreading the logic.